platform/openide.util.ui.svg/src/org/openide/util/svg/DenyingElementLoader.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.openide.util.svg;
+
+import com.github.weisj.jsvg.attributes.AttributeParser;
+import com.github.weisj.jsvg.parser.ElementLoader;
+import com.github.weisj.jsvg.parser.ParsedDocument;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+class DenyingElementLoader implements ElementLoader {
+    private final Set<String> attemptedExternalURLsLoaded = new LinkedHashSet<>();
+
+    public Set<String> getAttemptedExternalURLsLoaded() {
+        return Collections.unmodifiableSet(attemptedExternalURLsLoaded);
+    }
+
+    @Override
+    public <T> T loadElement(Class<T> type, String value,
+            ParsedDocument document, AttributeParser attributeParser)
+    {
+        /* Same logic as in com.github.weisj.jsvg.parser.DefaultElementLoader for the
+        AllowExternalResources.DENY case, but gathering up the attempted externally loaded URLs so
+        we can make the whole loading operation fail and make testLoadImageWithExternalUseXlinkHref
+        pass. */
+        String url = attributeParser.parseUrl(value);
+        if (url == null) {
+            return null;
+        }
+        if (url.contains("#")) {
+            String[] parts = url.split("#", 2);
+            String name = parts[0];
+            if (!name.isEmpty()) {
+                attemptedExternalURLsLoaded.add(value);
+                return null;
+            }
+            return document.getElementById(type, parts[1]);
+        } else {
+            return document.getElementById(type, url);
```

Comment:
From the first part of the logic and the name of the class I understand, that this should deny loading from external resources. If so, I think this else branch should read:

```suggestion
            attemptedExternalURLsLoaded.add(value);
            return null;
```

The else part is invoked when no `#` is found in the url. The url is thus a pure external reference and not an internal one.

Reply:
The code here was intended to do exactly the same as https://github.com/weisJ/jsvg/blob/v1.6.1-rc1/jsvg/src/main/java/com/github/weisj/jsvg/parser/DefaultElementLoader.java , without changing the logic. But it's good idea to convince ourselves that no external loading is happening here.

I think the variable name "url" might be misleading; often it's just strings such as "white", "black", and "rgb(204,102,53)" (confirmed via System.out.println calls.)  The "parseUrl" method returns the original string except strips surrounding "url(...)" if present.

Inspecting ParsedDocument.getElementById, I can confirm it doesn't do any external I/O; it just looks up the parameter in a map (which is supposed to be an element ID, not a full URL).

Reply:
Ah sorry, overlooked the fact, that resulted just in a lookup. To much assumption, to little reading on my side. Sorry for the noise.

Reply:
No I agree, it wasn't obvious at all. Thanks for reviewing!